### PR TITLE
Fixed resetting of players + added a listener system (on goal)

### PR DESCRIPTION
**Pull request by BeeHiveJava**

### CODE REVIEW

File: Soccar [Physics]/src/nl/soccar/physics/WorldObject.java [hunk 1]
```diff
@@ -46,4 +46,6 @@
      */
     float getDegree();
 
+    boolean isResetting();
```
Comment (ghost) [inline Soccar [Physics]/src/nl/soccar/physics/WorldObject.java:49]: MINOR Document this public method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [Physics]/src/nl/soccar/physics/models/WheelPhysics.java [hunk 2]
```diff
@@ -172,49 +215,8 @@ public void setDesiredSpeed(ThrottleAction throttleAction) {
...
-    public void reset() {
-        setPosition(originalPos.x, originalPos.y, 0, 0, 0, 0);
-    }
-
-    @Override
-    public void setPosition(float x, float y, float degree, float linearVelocityX, float linearVelocityY, float angularVelocity) {
-        body.setLinearVelocity(new Vec2(0.0F, 0.0F));
-        body.setAngularVelocity(0.0F);
-        body.getPosition().set(x, y);
+    private void setAngle(float angle) {
```
Comment (ghost) [inline Soccar [Physics]/src/nl/soccar/physics/models/WheelPhysics.java:218]: MAJOR Remove this unused private "setAngle" method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUnusedPrivateMethod)

File: Soccar [Physics]/src/nl/soccar/physics/GameEngine.java [hunk 3]
```diff
@@ -72,6 +74,22 @@ public void stop() {
...
 
+    public void addListener(GameEventListener listener) {
+        Objects.requireNonNull(listener);
+
+        synchronized (listeners) {
+            listeners.add(listener);
+        }
+    }
+
+    public void removeListener(GameEventListener listener) {
```
Comment (ghost) [inline Soccar [Physics]/src/nl/soccar/physics/GameEngine.java:85]: MINOR Document this public method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [Physics]/src/nl/soccar/physics/GameEngine.java [hunk 4]
```diff
@@ -23,12 +19,16 @@
 public final class GameEngine {
 
     private final World world;
+
+    private final Session session;
+    private final Room room;
```
Comment (ghost) [inline Soccar [Physics]/src/nl/soccar/physics/GameEngine.java:24]: MAJOR Remove this unused "room" private field. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1068)

File: Soccar [Physics]/src/nl/soccar/physics/GameEngine.java [hunk 5]
```diff
@@ -37,8 +37,10 @@
      *
      * @param game
      */
-    public GameEngine(Game game) {
-        this.game = game;
+    public GameEngine(Session session) {
```
Comment (ghost) [inline Soccar [Physics]/src/nl/soccar/physics/GameEngine.java:40]: MINOR Document the parameter(s): session [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [Physics]/src/nl/soccar/physics/GameEngine.java [hunk 6]
```diff
@@ -72,6 +74,22 @@ public void stop() {
         timer.cancel();
     }
 
+    public void addListener(GameEventListener listener) {
```
Comment (ghost) [inline Soccar [Physics]/src/nl/soccar/physics/GameEngine.java:77]: MINOR Document this public method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [Physics]/src/nl/soccar/physics/GameEngine.java [hunk 7]
```diff
@@ -176,6 +188,14 @@ public void removeWorldObject(WorldObject object) {
         }
     }
 
+    public void resetWorldObjects() {
```
Comment (ghost) [inline Soccar [Physics]/src/nl/soccar/physics/GameEngine.java:191]: MINOR Document this public method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

File: Soccar [Physics]/src/nl/soccar/physics/listener/GameEventListener.java [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+import nl.soccar.library.Ball;
+import nl.soccar.library.Session;
+import nl.soccar.library.enumeration.EventType;
+import nl.soccar.physics.GameEngine;
+
+/**
+ * @author PTS34A
+ */
+public interface GameEventListener {
```
Comment (ghost) [inline Soccar [Physics]/src/nl/soccar/physics/listener/GameEventListener.java:11]: MINOR Annotate the "GameEventListener" interface with the @FunctionalInterface annotation (sonar.java.source not set. Assuming 8 or greater.) [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1609)

File: Soccar [Physics]/src/nl/soccar/physics/listener/GameEventListener.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+
+    void onBallInGoal(GameEngine engine, Session session, Ball ball, EventType type);
```
Comment (ghost) [inline Soccar [Physics]/src/nl/soccar/physics/listener/GameEventListener.java:13]: MINOR Document this public method. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by ghost** [comment]
SonarQube analysis reported 10 issues
* MAJOR 2 major
* MINOR 8 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [GameEngine.java#L111](https://github.com/PTS3-S34A/Physics/blob/a00c29a1d498ae0c4a24efa734fe80adfdd23ff4/Soccar [Physics]/src/nl/soccar/physics/GameEngine.java#L111): Replace this "switch" statement by "if" statements to increase readability. [rule](http://194.135.91.67:9000/coding_rules#rule_key=squid%3AS1301)

**Review by Ranray — APPROVED** [review]

**Pull request merged**